lib/cartopy/io/srtm.py

```diff
@@ -246,29 +249,68 @@ def srtm(lon, lat):
     return SRTM3Source().single_tile(lon, lat)
 
 
-def add_shading(elevation, azimuth, altitude):
+def add_shading(located_image, azimuth=315, altitude=45, hscale=111120,
+                zscale=1):
     """Adds shading to SRTM elevation data, using azimuth and altitude
     of the sun.
 
-    :type elevation: numpy.ndarray
-    :param elevation: SRTM elevation data (in meters)
+    Function assumes x,y,z in units of downloaded SRTM [degree,degree,meters],
+    so default horizontal scale is 111120, which is good for equatorial
+    regions, and adequate elsewhere.
+
+    Azimuth degrees clockwise from North. Altitude is angle above horizon.
+
+    Function modeled after GDAL script (gdaldem hillshade):
+    geospatialpython.com/2013/12/python-and-elevation-data-creating.html
+
+    :type located_image: LocatedImage instance
+    :param located_image: SRTM elevation data (in meters)
     :type azimuth: float
     :param azimuth: azimuth of the Sun (in degrees)
     :type altitude: float
     :param altitude: altitude of the Sun (in degrees)
+    :type hscale: float
+    :param hscale: ratio of vertical to horizontal units (111120 m/degree)
+    :type zscale: float
+    :param zscale: vertical exaggeration multiplier (1= no exaggeration)
 
     :rtype: numpy.ndarray
-    :return: shaded SRTM relief map.
+    :return: shaded SRTM relief map
     """
+    elevation = located_image.image
+
+    xres = ((located_image.extent[1] - located_image.extent[0]) /
+            elevation.shape[1])
+    yres = ((located_image.extent[3] - located_image.extent[2]) /
+            elevation.shape[0])
+
     azimuth = np.deg2rad(azimuth)
     altitude = np.deg2rad(altitude)
-    x, y = np.gradient(elevation)
-    slope = np.pi/2. - np.arctan(np.sqrt(x*x + y*y))
-    # -x here because of pixel orders in the SRTM tile
-    aspect = np.arctan2(-x, y)
-    shaded = np.sin(altitude) * np.sin(slope)\
-        + np.cos(altitude) * np.cos(slope)\
-        * np.cos((azimuth - np.pi/2.) - aspect)
+
+    window = []
+    for row in range(3):
+        for col in range(3):
+            window.append(elevation[row:(row + elevation.shape[0] - 2),
+                                    col:(col + elevation.shape[1] - 2)])
+
+    # Process each cell (average gradients)
+    x = (((zscale * window[0] + zscale * window[3] + zscale * window[3] +
```

Comment:
I think this might be done more easily with a skimage kernel.

For example:

```
skimage.filters.rank.sum(elevation, [[1, 1, 1], [1, 0, 1], [1, 1, 1,]])
```

Reply:
This should probably be window[4], not window[3], right?

Reply:
Makes sense to factor out the zscale term.

Reply:
Maybe instead of adding skimage as a dependency, we can simply make use of ``np.gradient``.